lib/Backend/NativeEntryPointData.cpp

```diff
@@ -0,0 +1,568 @@
+#endif
+
+    // if scriptContext is shutting down, no need to free that native code
+    if (!isShutdown && this->nativeAddress != nullptr)
+    {
+        // In the debugger case, we might call cleanup after the native code gen that
+        // allocated this entry point has already shutdown. In that case, the validation
+        // check below should fail and we should not try to free this entry point
+        // since it's already been freed
+        Assert(this->validationCookie != nullptr);
+        if (this->validationCookie == scriptContext->GetNativeCodeGenerator())
+        {
+            FreeNativeCodeGenAllocation(scriptContext, this->nativeAddress, this->thunkAddress);
+        }
+    }
+
+#ifdef PERF_COUNTERS
+    PERF_COUNTER_SUB(Code, TotalNativeCodeSize, codeSize);
+    PERF_COUNTER_SUB(Code, FunctionNativeCodeSize,codeSize);
+    PERF_COUNTER_SUB(Code, DynamicNativeCodeSize, codeSize);
+#endif
+
+    this->nativeAddress = nullptr;
+    this->thunkAddress = nullptr;
+    this->codeSize = 0;
+}
+
+void
+NativeEntryPointData::SetTJNativeAddress(Js::JavascriptMethod nativeAddress, void * validationCookie)
+{
+    Assert(this->nativeAddress == nullptr);
+    Assert(this->validationCookie == nullptr);
+    this->nativeAddress = nativeAddress;
+    this->validationCookie = validationCookie;
+}
+
+void 
+NativeEntryPointData::SetTJCodeSize(ptrdiff_t codeSize)
+{
+    Assert(this->codeSize == 0);
+    this->codeSize = codeSize;
+}
+
+void 
+NativeEntryPointData::AddWeakFuncRef(RecyclerWeakReference<FunctionBody> *weakFuncRef, Recycler *recycler)
+{
+    NativeEntryPointData::WeakFuncRefSet * weakFuncRefSet = this->weakFuncRefSet;
+    if (weakFuncRefSet == nullptr)
+    {
+        weakFuncRefSet = RecyclerNew(recycler, WeakFuncRefSet, recycler);
+        this->weakFuncRefSet = weakFuncRefSet;
+    }
+    weakFuncRefSet->AddNew(weakFuncRef);
+}
+
+EntryPointPolymorphicInlineCacheInfo * 
+NativeEntryPointData::EnsurePolymorphicInlineCacheInfo(Recycler * recycler, FunctionBody * functionBody)
+{
+    if (!polymorphicInlineCacheInfo)
+    {
+        polymorphicInlineCacheInfo = RecyclerNew(recycler, EntryPointPolymorphicInlineCacheInfo, functionBody);
+    }
+    return polymorphicInlineCacheInfo;
+}
+
+void 
+NativeEntryPointData::RegisterConstructorCache(Js::ConstructorCache* constructorCache, Recycler* recycler)
+{
+    Assert(constructorCache != nullptr);
+
+    if (!this->constructorCaches)
+    {
+        this->constructorCaches = RecyclerNew(recycler, ConstructorCacheList, recycler);
+    }
+
+    this->constructorCaches->Prepend(constructorCache);
+}
+
+void
+NativeEntryPointData::PinTypeRefs(Recycler * recycler, size_t jitPinnedTypeRefCount, void ** jitPinnedTypeRefs)
+{
+    this->runtimeTypeRefs = RecyclerNewArray(recycler, Field(void*), jitPinnedTypeRefCount + 1);
+    //js_memcpy_s(this->runtimeTypeRefs, jitPinnedTypeRefCount * sizeof(void*), jitPinnedTypeRefs, jitPinnedTypeRefCount * sizeof(void*));
```

Comment:
Why does the memcpy not work here? In any case, remove commented code?

Reply:
It was commented before, and so I copied.
I think the reason memcpy doesn't work because we need writebarrier triggers for xplat.
Added a comment about that and a TODO to switch to Memory::CopyArray.